core/src/main/java/org/elasticsearch/discovery/zen/publish/PendingClusterStatesQueue.java

```diff
@@ -164,12 +164,17 @@ public synchronized void markAsProcessed(ClusterState state) {
                             currentMaster
                     );
                 }
-            } else if (state.supersedes(pendingState) && pendingContext.committed()) {
+            } else if (state.version() >= pendingState.version()) {
+                assert state.supersedes(pendingState) || (
```

Comment:
is we're here, then the masters are equal. In this case supersedes == false, means version equality (because it's can't be lower. Also state.nodes().getMasterNodeId() is never null (and assigned to currentMaster.getId())) . I'm not sure what we're asserting than. Also note that it disables the next if clause, checking for uuid equality,  so maybe we want to do the next if clause first?